### DISCUSSION TITLE
How much does MDN want to document Wasm (/wat) itself?

### DISCUSSION THREAD

**Discussion by ghost** [mdn/mdn-community#558] [Content]
I fully believe that MDN should document the WebAssembly APIs/interfaces/etc. for JavaScript; no question there.

Instead, I am concerned with pages like these:
* [/docs/WebAssembly/Understanding_the_text_format](https://developer.mozilla.org/docs/WebAssembly/Understanding_the_text_format)
* [/docs/WebAssembly/Text_format_to_wasm](https://developer.mozilla.org/docs/WebAssembly/Text_format_to_wasm)
* [/docs/WebAssembly/Reference](https://developer.mozilla.org/docs/WebAssembly/Reference)

As MDN describes itself as "web docs," and documents HTML and CSS, it establishes some precedent for a decision to document Wasm too.
Understandably, considering how WebAssembly is much newer than the prior technologies, and how much more complicated it is, comparatively, the documentation on MDN dedicated to Wasm is far less exhaustive and sophisticated than that of the existing JS/CSS/HTML.

How much we want to document? How thorough should examples be? How exhaustive should documentation be? Should there be a `/learn` section for Wasm? Should every single instruction be documented?

The last question especially sounds like overkill for MDN; why not leave this to the specification and others? There are hundreds of instructions, and hundreds more to come - to comprehensively cover them all, and covering their semantics along with creating examples could be difficult, although possibly only as difficult as it was for the JS/HTML/CSS `/Reference`s.

I really believe that we could benefit from explicitly stating what we want here, giving us a good idea as to how to guide further development of the documentation.

**Comment by hamishwillee** [comment]
I agree we need a stated goal. Minimally we need to cover JavaScript interfaces. To some extent that means we need to provide guide material because it is not easy to understand the semantics of those without the bigger picture of what is happening in WASM.
I don't think we should contemplate "Learn" for WASM - who would be the audience for that?

@MendyBerger

**Reply by ghost** [reply]
(Fwiw, I agree on the `/learn` part! I can't imagine that providing much value, especially contrasted against the price)

**Comment by teoli2003** [comment]
cc/ @ms2ger that will have a good insight here!

**Comment by wingo** [comment]
A personal opinion from someone working on WebAssembly: MDN should avoid documenting WebAssembly itself.  However I agree with @hamishwillee that it would be very useful to have better user-focussed documentation on the JS interface (https://webassembly.github.io/spec/js-api/index.html#webassembly-namespace).  I would not mention the text format; instead I would focus on byte arrays, either in JS or fetched from a server.  The various secure-origin and MIME restrictions are worth documenting as well, for fetch origins.

**Comment by MendyBerger** [comment]
I'll share my thoughts here since I wrote most of the `wat` reference.

#### Value:
I'm not in a position to say if it's worth doing, can only say that I started to write them because I was looking for accessible docs not written for compiler authors, don't know if I'm unique in that way but I suspect not.

I think that it can be really valuable for developers if we do document the `wat` format **even if it's not as thorough as the other docs**. Having the ability to see a list of available instructions with a simple description and a simple example can help people understand what wasm is or isn't. Understanding this has certainly helped me understand whats happening the to the code I write (I write Rust on wasm full time).

#### Not as thorough:
Granted, it's much less exhaustive and sophisticated but that's because it's new, I'm not very good at writing text/docs, so I created a ready page with basic descriptions and runnable examples, the description might not mention all the semantics but that can now be added in easily, it's not exhaustive but I think it's acceptable. (I don't think that the other docs are complete when first added, they're meant to get better over time, similar to a wiki)

#### Required work / large spec:
The `wasm` spec is much much smaller than the JS or CSS specs, it's actually pretty small, every instruction from the core spec is already documented on MDN in the `wat` reference. It's going to take orders of magnitude less work to bring them up to the quality of the other docs, albeit with much less return on investment.

There aren't that many instructions, lots if not most of them are the same instruction for different types. For example, there are 6 division instructions for types: `i32`, `u32`, `f32`, `i64`, `u64`, `f64`. I don't think we need to document each if them separately, they're effectively doing the same thing from the developer's point of view.


#### Explicitly defined goal:
A clear decision would definitely be valuable.

(If the admins decide to remove them, can I get permission to take the content and put them somewhere else? Really don't want all that work to be wasted)

Thoughts?

**Reply by ghost** [reply]
> #### Value:
> 
> I'm not in a position to say if it's worth doing, can only say that I started to write them because I was looking for accessible docs not written for compiler authors, don't know if I'm unique in that way but I suspect not.

Although, curiously, we actually do note that it is more useful for someone who is implementing a Wasm compiler, in [/docs/WebAssembly/Understanding_the_text_format](https://developer.mozilla.org/docs/WebAssembly/Understanding_the_text_format), that is, that knowledge of the wasm text format is very niche. (It was not clear as to which compiler you were referring to, regarding language->wasm, vs. wasm->native)

> #### Not as thorough:
> 
> Granted, it's much less exhaustive and sophisticated but that's because it's new, I'm not very good at writing text/docs, so I created a ready page with basic descriptions and runnable examples, the description might not mention all the semantics but that can now be added in easily, it's not exhaustive but I think it's acceptable. (I don't think that the other docs are complete when first added, they're meant to get better over time, similar to a wiki)

I do not see a this aspect to be problematic. I am interested in updating the documents to reflect specification extensions that have since been merged, adding new examples, expanding documentation, etc. I am particularly interested in having a better motivation and understanding for how these docs should be written, guiding their next iteration(s).

> #### Required work / large spec:
> 
> The `wasm` spec is much much smaller than the JS or CSS specs, it's actually pretty small, every instruction from the core spec is already documented on MDN in the `wat` reference. It's going to take orders of magnitude less work to bring them up to the quality of the other docs...
> 
> There aren't that many instructions, lots if not most of them are the same instruction for different types. For example, there are 6 division instructions for types: `i32`, `u32`, `f32`, `i64`, `u64`, `f64`. I don't think we need to document each if them separately, they're effectively doing the same thing from the developer's point of view.

Ah, I had not considered coalescing instruction documentation, but that is actually how it already is written. I had in mind the `/Refererence` page. The reference has a link to an unwritten page about the (fixed-width) SIMD instructions, yet there are over 200 instructions instructed by that proposal alone (granted, this is an outlier among proposals, besides the [length-agnostic SIMD proposal](https://github.com/WebAssembly/flexible-vectors)). This is a very valid point, but there is also the complexity of explaining all of these new features. Looking towards the finished proposals and future proposals for examples, these all begin to require far more low-level details, to a point that I am not certain as to how they should be described without a sharp idea of who the audience is in mind.

Some examples of independent concepts that I believe would need to be covered which most web developers would likely be unfamiliar with, would be byte endianness (generalizations into shuffling larger integer sizes) [[1]](https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#swizzling-using-variable-indices), sign/zero extension [[2]](https://github.com/WebAssembly/sign-extension-ops), tail calling [3](https://github.com/WebAssembly/tail-call), possibly WebIDL (for the interface types / component-model proposal), branch prediction [[3]](https://github.com/WebAssembly/branch-hinting), atomic/shared memory (`SharedArrayBuffer`/`Atomics` are not frequently used) [[4]](https://github.com/webassembly/threads), memory alignment, time variance of code execution [[5]](https://github.com/WebAssembly/constant-time), among so many other complicated niche topics.

(This is merely to drive in the point that furthering the documentation in an understandable manner for new topics may actually be more costly than one may immediately/intuitively suspect!)

Now, there is the alternative perspective that you reference: (low-level?) developers who have familiarity with languages that compile to WebAssembly, such as C++, Rust, Zig, C, and such languages. Proficient programmers in a lower-level field or simply with lower-level functionality are far more likely to have attained some degree of familiarity and a mental model of general ISAs, which builds a useful background which can be easily extended to Wasm.

But again, who is MDN's Wasm demographic? What audience are we aiming for?

> (If the admins decide to remove them, can I get permission to take the content and put them somewhere else? Really don't want all that work to be wasted)

At a minimum, I definitely wouldn't like to see them removed! :)

And overall, feel free to point out any of my fears that may be over-exaggerated!

**Reply by MendyBerger** [reply]
> The reference has a link to an unwritten page about the (fixed-width) SIMD instructions

Right, that SIMD link should probably be removed, most of it can be folded into existing numeric pages. I've started to look into how SIMD should be structred, but I don't have a solution yet. (feel free to propose something)

> Some examples of independent concepts that I believe would need to be covered

1. Some of the complex topics already have MDN pages like [Endianness](https://developer.mozilla.org/en-US/docs/Glossary/Endianness), and for others a link to Wikipedia might be enough.
2. What if we don't provide these pages right away, someone can come later and add it in, and if that doesn't happen, is there no value there anymore? I don't think that the docs isn't useful if we don't cover memory alignment, if it's covered then great! But otherwise, I'd argue that it's still useful.

> (low-level?)

Not strictly low level developers, there are high level languages that compile to wasm like assembly script, c#, and python. And developers of those languages might find these docs useful as well, but low level developers will certainly find it easier to grasp.

> But again, who is MDN's Wasm demographic? What audience are we aiming for?

I'd say it's similar to the overall MDN demographic, ranging from basic novices with only a rudimentary knowledge about those technologies, to experts with a deep understanding of those topics, but does not include people with no clue how computers work, or on the other side compiler authors. (the wasm novice is probably a lot more advanced then the general web novice though)

**Reply by hamishwillee** [reply]
FWIW I'm a niche audience, but when I started the JS WASM exception handling docs I found our existing guide, and (to a lesser extent) the broader WASM reference invaluable as a starting point. More generally, the interactive examples make it easy to see what is supported and "what is going on". I used this as the basis for working out how to write examples that made sense for the JS docs. It does need to be built-out and made consistent.

If this were a blank page we'd be asking "is it worth the cost of writing" and "is it worth the cost of maintaining". For the first case, this is essentially happening for free - thanks @MendyBerger. So at least to me, absolutely this is worth keeping and extending in its current form. 
With respect to maintenance, only time will tell. However what I'm seeing is a relatively stable foundation in WASM that I do not _think_ will age badly. 

The reason I am opposed to "learn" is that it is another layer over our more terse guide that is not worth investing in.

**Comment by sideshowbarker** [comment]
@Kylebrown9 It’d be great to have your feedback on this — and also great if you might be able to think of others who could give some feedback (and maybe consider Cc’ing them on this)

**Reply by esoterra** [reply]
I'll second what people have already been saying that detailed documentation can help people understand what WASM is and what it can do making it more broadly useful than it seems. I think we should cover things like the instructions and especially control flow semantics, which I think people might find unintuitive at first, to whatever extent is possible.

Passed it along to the [Bytecode Alliance](https://bytecodealliance.org/) [Zulip group](https://bytecodealliance.zulipchat.com/), some of those folks might have suggestions.
